src/Trace.ts

```diff
@@ -0,0 +1,219 @@
+   */
+  getContext(): TraceContext {
+    return Object.assign({}, this._traceContext, { tags: JSON.parse(JSON.stringify(this.tags)) })
+  }
+
+  /**
+   * Creates and returns new child span of the current span and changes the span service name
+   * @param service the name of the service of the new child span
+   */
+  getChildSpan(service: string): Trace {
+    if (this._traceContext.finishTimestamp) throw new Error('Finished trace cannot have a child span')
+    let inputTraceContext: IEventTrace = <IEventTrace>this.getContext()
+    if (!(inputTraceContext.traceId && inputTraceContext.spanId) && !(inputTraceContext.service)) {
+      throw new Error('No Service or traceId or SpanId provided')
+    }
+    return new Trace(new EventTraceMetadata(Object.assign({},
+      inputTraceContext, {
+        service,
+        spanId: undefined,
+        parentSpanId: inputTraceContext.spanId
+      })))
+  }
+
+  /**
+   * Injects trace context into a carrier with optional path.
+   * @param carrier any kind of message or other object with keys of type String.
+   * @param injectOptions type and path of the carrier. Type is not implemented yet. Path is the path to the trace context.
+   */
+  injectContextToMessage(carrier: { [key: string]: any }, injectOptions: IContextOptions = {}): Promise<{ [key: string]: any }> {
+    let result = carrier
+    let { path } = injectOptions // type not implemented yet
+    if (carrier instanceof EventMessage || (('metadata' in carrier))) path = 'metadata.trace'
+    else if (('trace' in carrier)) path = 'trace'
+    else if (carrier instanceof EventTraceMetadata) result.metadata.trace = this._traceContext
+    if (path) {
+      try {
+        let pathArray: string[] = path.split('.')
+        for (let i = 0; i < pathArray.length - 1; i++) {
+          if (!result[pathArray[i]]) {
+            if (i < pathArray.length) {
+              let o: any = {}
+              o[pathArray[i + 1]] = {}
+              result[pathArray[i]] = o
+            }
+          }
+          result = result[pathArray[i]]
+        }
+      } catch (e) {
+        throw e
+      }
+    }
+    result.trace = this._traceContext
+    return Promise.resolve(carrier)
+  }
+
+  /**
+   * Sets tags to the current span. If child span is created, the tags are passed on.
+   * @param tags key value pairs of tags. Tags can be changed on different child spans
+   */
+  setTags(tags: { [key: string]: any }): this {
+    let newContext: IEventTrace = new EventTraceMetadata(this.getContext())
+    if (!newContext.tags) {
+      newContext.tags = tags
+    } else {
+      for (let key in tags) {
+        newContext.tags[key] = tags[key]
+      }
+    }
+    this._traceContext = Object.freeze(new EventTraceMetadata(newContext))
+    this._updateContext()
+    return this
+  }
+
+  /**
+   * Finishes the trace by adding finish timestamp to the current span.
+   * @param finishTimestamp optional parameter for the finish time. If omitted, current time is used.
```

Comment:
Why would the trace object have the traceContext?

I believe only the Span objects would have the traceContext.